adlchecker/src/main/java/com/nedap/archie/adlchecker/TerminologyContentGenerator.java

```diff
@@ -0,0 +1,167 @@
+        Map<String, Map<String, ArchetypeTerm>> termDefinitions = resultingArchetype.getTerminology().getTermDefinitions();
+        for(String language: termDefinitions.keySet()) {
+            Map<String, ArchetypeTerm> stringArchetypeTermMap = termDefinitions.get(language);
+            Map<String, ArchetypeTerm> sortedArchetypeTermMap = new TreeMap<>(new Comparator<String>() {
+                @Override
+                public int compare(String o1, String o2) {
+                    String code1 = o1.substring(0, 2);
+                    String code2 = o2.substring(0, 2);
+                    if(code1.equalsIgnoreCase(code2)) {
+                        int number1 = Integer.parseInt(o1.substring(2));
+                        int number2 = Integer.parseInt(o2.substring(2));
+                        return number1 - number2;
+                    } else {
+                        return code2.compareTo(code1);
+                    }
+                }
+            });//TODO: comparator!
+            sortedArchetypeTermMap.putAll(stringArchetypeTermMap);
+            termDefinitions.put(language, sortedArchetypeTermMap);
+
+        }
+    }
+
+    private void walkArchetype(String sourceFile, Archetype archetype, Archetype resultingArchetype) {
+        Stack<CObject> workList = new Stack<>();
+        workList.push(archetype.getDefinition());
+        List<CObject> cObjects = new ArrayList<>();
+        while(!workList.empty()) {
+            CObject next = workList.pop();
+            if(!terminologyHasCode(archetype, next)) {
+                cObjects.add(next);
+                addCodeToTerminology(sourceFile, resultingArchetype, next);
+            }
+            for(CAttribute attribute:next.getAttributes()) {
+                for(CObject child:attribute.getChildren()) {
+                    workList.push(child);
+                }
+            }
+        }
+    }
+
+    private void addCodeToTerminology(String sourceFile, Archetype resultingArchetype, CObject next) {
+        String text = getCommentName(sourceFile, next.getNodeId());
+        if(text == null) {
+            text = "X";
+        }
+        String description = text; //TODO: grep comment from sourcefile!
+        Map<String, Map<String, ArchetypeTerm>> termDefinitions = resultingArchetype.getTerminology().getTermDefinitions();
+        for(String language: termDefinitions.keySet()) {
+            if(termDefinitions.get(language).get(next.getNodeId()) == null) {
+                ArchetypeTerm newCode = new ArchetypeTerm();
+                newCode.setCode(next.getNodeId());
+                newCode.setText(text);
+                newCode.setDescription(description);
+                termDefinitions.get(language).put(next.getNodeId(), newCode);
+            }
+        }
+    }
+
+    private String getCommentName(String sourceFile, String nodeId) {
+        String[] lines = sourceFile.split("\n");
+        for(String line:lines) {
```

Comment:
Does this retrieve all the comments in the archetype every time when you need one specific comment?

Reply:
yes it does. Works fine for now, I'll add a comment